From 69e5a07c450bd3dff623554f068edd853dd989c6 Mon Sep 17 00:00:00 2001
From: oreiche <oliver.reiche@gmail.com>
Date: Fri, 23 May 2014 17:31:24 +0200
Subject: [PATCH 1/2] SystemUI: Extend SwipeHelper

I was not happy with long press on notifications and recents
to open apps in floating mode. The long press duration kinda
breaks the UI interaction flow, which prevented me from
actually using floating windows. Therefore, I've created a
private modification, which enables floating mode for
notifications and recents by swiping to the left. This
greatly improves usability without strongly deviating from
stock UI experience, in strict accordance with PA's
philosophy.  Changes are surgically implemented to keep the
probability of merge conflicts rather low. The code was
successfully tested on mako, hammerhead, tilapia, and manta.
Let the team discuss and decide whether or not this is a
worthy extension to PA.

Changes are split into two commits (PT1 & PT2).

PT1:
- Extending SwipeHelper:
SwipeHelper can now distinguish between directions LEFT,
RIGHT, UP, DOWN and can fire an abstract trigger 'event'. A
trigger direction can be specified and when a swipe in this
direction occurs dismissable children will not be dismissed,
instead they will be snapped back.  Whenever a child is
swiped far or fast enough in that direction an event is
triggered and the onChildTriggerd() callback is issued. This
callback was added to the SwipeHelper.Callback interface and
therefore requires all its implementations to realize this
method, such as probably Hover, which is sadly not open
source (yet?).

PS7: Added Hover support

TODO: Edit commit message ;)

Change-Id: Ida06195913f022463d29ca22092e81fb8993ad50
---
 .../SystemUI/src/com/android/systemui/SwipeHelper.java     | 14 ++++++++++++++
 .../systemui/recent/RecentsHorizontalScrollView.java       |  3 +++
 .../android/systemui/recent/RecentsVerticalScrollView.java |  3 +++
 .../systemui/statusbar/notification/HoverLayout.java       |  4 ++++
 .../systemui/statusbar/notification/PeekLayout.java        |  4 ++++
 .../systemui/statusbar/policy/HeadsUpNotificationView.java |  4 ++++
 .../systemui/statusbar/policy/NotificationRowLayout.java   |  3 +++
 7 files changed, 35 insertions(+)

diff --git a/packages/SystemUI/src/com/android/systemui/SwipeHelper.java b/packages/SystemUI/src/com/android/systemui/SwipeHelper.java
index e7c0d02..2e876df 100644
--- a/packages/SystemUI/src/com/android/systemui/SwipeHelper.java
+++ b/packages/SystemUI/src/com/android/systemui/SwipeHelper.java
@@ -75,6 +75,12 @@ public class SwipeHelper implements Gefingerpoken {
     private Runnable mWatchLongPress;
     private long mLongPressTimeout;
 
+    private int mTriggerDirection;
+    public static final int LEFT = 0;
+    public static final int RIGHT = 1;
+    public static final int UP = 2;
+    public static final int DOWN = 3;
+
     public SwipeHelper(int swipeDirection, Callback callback, float densityScale,
             float pagingTouchSlop) {
         mCallback = callback;
@@ -297,6 +303,7 @@ public class SwipeHelper implements Gefingerpoken {
         anim.addListener(new AnimatorListenerAdapter() {
             public void onAnimationEnd(Animator animation) {
                 mCallback.onChildDismissed(view);
+                mCallback.onChildDismissed(view, mTriggerDirection);
                 animView.setLayerType(View.LAYER_TYPE_NONE, null);
             }
         });
@@ -348,6 +355,11 @@ public class SwipeHelper implements Gefingerpoken {
             case MotionEvent.ACTION_MOVE:
                 if (mCurrView != null) {
                     float delta = getPos(ev) - mInitialTouchPos;
+
+                    mTriggerDirection = delta < 0 ?
+                        (mSwipeDirection == X ? LEFT : UP) :
+                        (mSwipeDirection == X ? RIGHT : DOWN);
+
                     // don't let items that can't be dismissed be dragged more than
                     // maxScrollDistance
                     if (CONSTRAIN_SWIPE && !mCallback.canChildBeDismissed(mCurrView)) {
@@ -408,6 +420,8 @@ public class SwipeHelper implements Gefingerpoken {
 
         void onChildDismissed(View v);
 
+        void onChildDismissed(View v, int direction);
+
         void onDragCancelled(View v);
     }
 }
diff --git a/packages/SystemUI/src/com/android/systemui/recent/RecentsHorizontalScrollView.java b/packages/SystemUI/src/com/android/systemui/recent/RecentsHorizontalScrollView.java
index e1aac9f..b78ac05 100644
--- a/packages/SystemUI/src/com/android/systemui/recent/RecentsHorizontalScrollView.java
+++ b/packages/SystemUI/src/com/android/systemui/recent/RecentsHorizontalScrollView.java
@@ -255,6 +255,9 @@ public class RecentsHorizontalScrollView extends HorizontalScrollView
         contentView.setTranslationY(0);
     }
 
+    public void onChildDismissed(View v, int direction) {
+    }
+
     public void onBeginDrag(View v) {
         // We do this so the underlying ScrollView knows that it won't get
         // the chance to intercept events anymore
diff --git a/packages/SystemUI/src/com/android/systemui/recent/RecentsVerticalScrollView.java b/packages/SystemUI/src/com/android/systemui/recent/RecentsVerticalScrollView.java
index 05d7847..7b810d4 100644
--- a/packages/SystemUI/src/com/android/systemui/recent/RecentsVerticalScrollView.java
+++ b/packages/SystemUI/src/com/android/systemui/recent/RecentsVerticalScrollView.java
@@ -262,6 +262,9 @@ public class RecentsVerticalScrollView extends ScrollView
         contentView.setTranslationX(0);
     }
 
+    public void onChildDismissed(View v, int direction) {
+    }
+
     public void onBeginDrag(View v) {
         // We do this so the underlying ScrollView knows that it won't get
         // the chance to intercept events anymore
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/notification/HoverLayout.java b/packages/SystemUI/src/com/android/systemui/statusbar/notification/HoverLayout.java
index 6d16143..daed08b 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/notification/HoverLayout.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/notification/HoverLayout.java
@@ -197,6 +197,10 @@ public class HoverLayout extends RelativeLayout implements ExpandHelper.Callback
         }
 
         @Override
+        public void onChildDismissed(View v, int direction) {
+        }
+
+        @Override
         public void onChildDismissed(View v) {
             mHover.clearHandlerCallbacks();
             mHover.setAnimatingVisibility(false);
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/notification/PeekLayout.java b/packages/SystemUI/src/com/android/systemui/statusbar/notification/PeekLayout.java
index f626295..892ff02 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/notification/PeekLayout.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/notification/PeekLayout.java
@@ -104,6 +104,10 @@ public class PeekLayout extends LinearLayout implements SwipeHelper.Callback {
     }
 
     @Override
+    public void onChildDismissed(View v, int direction) {
+    }
+
+    @Override
     public void onBeginDrag(View v) {
         mPeek.setAnimating(true);
     }
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/HeadsUpNotificationView.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/HeadsUpNotificationView.java
index 049d51e..1b44983 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/HeadsUpNotificationView.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/HeadsUpNotificationView.java
@@ -217,6 +217,10 @@ public class HeadsUpNotificationView extends FrameLayout implements SwipeHelper.
     }
 
     @Override
+    public void onChildDismissed(View v, int direction) {
+    }
+
+    @Override
     public void onBeginDrag(View v) {
     }
 
diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/NotificationRowLayout.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/NotificationRowLayout.java
index 259422d..86ac093 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/NotificationRowLayout.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/NotificationRowLayout.java
@@ -174,6 +174,9 @@ public class NotificationRowLayout
         }
     }
 
+    public void onChildDismissed(View v, int direction) {
+    }
+
     public void onBeginDrag(View v) {
         // We need to prevent the surrounding ScrollView from intercepting us now;
         // the scroll position will be locked while we swipe
-- 
1.9.1

